### PR TITLE
Support Unicode in chat & frag

### PR DESCRIPTION
**Pull request by HenryQuan**
This PR addresses https://github.com/WoWs-Builder-Team/minimap_renderer/issues/6 to support Chinese, Japanese and Korean languages in Chat. Now, the font is determined based on the message language. 

Two new dependencies are added, `langdetect` and `hanzidentifier`. Unfortunately, langdetect can detect Chinese as Korean, so the second library is required. 

However, there are several issues with this implementation.
- Huge font sizes (11MB for Chinese, 5MB for Korean and 6MB for Japanese)
- More memory usage (23MB in total)
- Language detection can be inaccurate

### CODE REVIEW

File: src/renderer/layers/chat.py [hunk 1]
```diff
@@ -142,3 +142,32 @@ def unpack_color(packed_value: int) -> tuple:
...
+        chat_message = message.message
+        language = detect(message.message) # this can detect Chinese as Korean
+        if has_chinese(chat_message):
+            return self._renderer.resman.load_font(
+                filename="warhelios_bold_zh.ttf", size=12
+            )
+    
+        if language == "ja":
+            return self._renderer.resman.load_font(
+                filename="warhelios_bold_ja.ttf", size=12
```
Comment (HenryQuan) [inline src/renderer/layers/chat.py:164]: This should be updated to jp instead ja.

**Comment by HenryQuan** [comment]
I have moved my function to resman instead of putting it under chat. There is also a minor dependency issue when I upgraded to Python 3.11. `lxml==4.9.1` doesn't work somehow, I had to use `lxml==4.9.4` locally. I didn't commit this change because I don't know if it breaks anything.

There is also one additional fix for this PR. I have supported rendering Chinese player names under frag. This is CN realm specific, so other servers aren't affected by this change.

For Chinese names to be rendered, I have to update `battle_controller` where I add the encoding under line 742 (12_11_1). I will also open an issue under replays_unpack.
```
name=player["name"].encode('ISO8859-1').decode('UTF-8'),
```

**Comment by wihn2021** [comment]
maybe no need to detect language, just let users add an argument, for example, "python -m render xxxx.wowsreplay --lang chinese", default argument is English. Then renderer loads different font files and ship infomation files

**Comment by HenryQuan** [comment]
@wihn2021 That is a good idea, but the chat may also contain Japanese or Korean at any moment. The Chinese font doesn't render all Japanese character or any Korean alphabet. Maybe, the flag can be something like `--unicode` to enable this feature or not. The renderer can determine the server of the replay which is used to render frags correctly in this PR.

**Comment by MeowthTim** [comment]
> For Chinese names to be rendered, I have to update `battle_controller` where I add the encoding under line 742 (12_11_1). I will also open an issue under replays_unpack.
> 
> name=player["name"].encode('ISO8859-1').decode('UTF-8'),
> ```

That is a very useful step for Chinese support. To use especilly for CN server without JP or KR, it just need a little change. The other thing I guess is translating the ship.json to Chinese which I would try to solve with the zh_lang.json in game. Thanks for your job.

**Comment by HenryQuan** [comment]
@MeowthTim I have the game language file in all supported languages. It is definitely possible to localise even further to support the Chinese server better.

**Comment by padtrack** [comment]
I put off addressing this because I was working on a rewrite that I burned out on. I'll merge this the next chance I have to work on personal projects.

**Comment by padtrack** [comment]
I am reviewing this on the train. Can you upload some test replays? I had some mixed language replays when I worked on this. The game does not load more than one font, but I think it would be nice to support all CJK characters and more together.

**Comment by padtrack** [comment]
> @wihn2021 That is a good idea, but the chat may also contain Japanese or Korean at any moment. The Chinese font doesn't render all Japanese character or any Korean alphabet. Maybe, the flag can be something like `--unicode` to enable this feature or not. The renderer can determine the [realm](https://github.com/WoWs-Builder-Team/minimap_renderer/blob/64a27ce6ef84ff1ccb22b34cbb512338bf2effcd/src/renderer/layers/frag.py#L29) of the replay which is used to render frags correctly in this PR.

Do you have example of JP characters the Chinese font doesn't support? I checked my notes from May 2023 and I saw Chinese font supported all random JP phrases I generated.

**Comment by HenryQuan** [comment]
@padtrack While working on this, I found out that WoWs provided different fonts for Korean, Japanese and Chinese.

I also have a feeling that the Chinese Font supports most Japanese characters, but I am unsure if this is a valid assumption. However, the Japanese Font doesn't support all Chinese characters. If we want to remove a Font, it could be the Japanese one. On the other hand, the Korean Font should be kept.

The upstream has also fixed some encoding issues regarding the clan name in https://github.com/Monstrofil/replays_unpack/issues/30. Maybe, @JustOneSummer dalao can provide some replays. I don't have any at hand now. We can have a training room replay at some time including all three languages.

**Comment by padtrack** [comment]
It's alright, I think I can probably find the old replays I had when I'm home. If you had any on hand it would have been convenient.

Screenshots from when I discussed with a JP speaker:
image

image

**Comment by padtrack** [comment]
@HenryQuan Can you enable `Allow edits from maintainers`? I also sent a message to what I think is your Discord.

**Pull request merged**